### PR TITLE
fix: fix bug when copying recursively with certain platforms

### LINKED ISSUE #812: Oras attach with plateform weird behavior when using oras cp
### What happened in your environment?

I have done this:

```
oras cp ghcr.io/oras-project/oras:v1.0.0-rc.1 --to-oci-layout ./local/oras:v1.0.0-rc.1 -v
```
```
oras attach --debug --oci-layout --artifact-type test.sbom ./local/oras@sha256:d1903141acefdeed52e34811d495f7569aba7b97d69c5a7ad22a3f213679b117 sbom 
```
```
oras discover --oci-layout ./local/oras:v1.0.0-rc.1 -v -o tree --platform linux/s390x
```

the output is

```
v1.0.0-rc.1
└── test.sbom
    └── sha256:f5a5798d17e52a0cf5094b835c45e87a138b1e041d945882b7ae396c0b8bda48
        └── org.opencontainers.artifact.created: "2023-02-14T10:22:14Z"
```
After that, I want to copy this to an azure repository with the following:

```
oras cp -r --debug --from-oci-layout -v ./local/oras:v1.0.0-rc.1 xxx.azurecr.io/artifactstest:testfororas
```
output:

```
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000] > Request method: "HEAD"                     
DEBU[0000] > Request headers:                           
DEBU[0000]    "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "401 Unauthorized"        
DEBU[0000] < Response headers:                          
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "X-Ms-Correlation-Request-Id": "ab852b21-cc32-4f64-9716-f3529afb081a" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "Www-Authenticate": "Bearer realm=\"https://xxx.azurecr.io/oauth2/token\",service=\"xxx.azurecr.io\",scope=\"repository:artifactstest:pull\"" 
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:23:52 GMT"   
DEBU[0000]    "Content-Length": "207"                   
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "Server": "openresty"                     
DEBU[0000] > Request URL: "https://xxx.azurecr.io/oauth2/token?scope=repository%3Aartifactstest%3Apull&service=xxx.azurecr.io" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "X-Ms-Ratelimit-Remaining-Calls-Per-Second": "166.65" 
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:23:52 GMT"   
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "X-Ms-Correlation-Request-Id": "432cc6a8-9de1-4de6-a32c-db116afe19c2" 
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000] > Request method: "HEAD"                     
DEBU[0000] > Request headers:                           
DEBU[0000]    "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "X-Ms-Client-Request-Id": ""              
DEBU[0000]    "X-Ms-Request-Id": "eeb4e549-d1a2-45ef-b92f-9c230ebb645e" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:23:52 GMT"   
DEBU[0000]    "Content-Type": "application/vnd.oci.image.index.v1+json" 
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Docker-Content-Digest": "sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000]    "Etag": "\"sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b\"" 
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Content-Length": "3138"                  
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "X-Ms-Correlation-Request-Id": "53bee9d4-6c62-4e6c-961a-092460a450eb" 
Exists  4169ef297659 application/vnd.oci.image.index.v1+json
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Accept": "application/vnd.oci.image.index.v1+json" 
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "401 Unauthorized"        
DEBU[0000] < Response headers:                          
DEBU[0000]    "Content-Length": "207"                   
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "Www-Authenticate": "Bearer realm=\"https://xxx.azurecr.io/oauth2/token\",service=\"xxx.azurecr.io\",scope=\"repository:artifactstest:pull\"" 
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:23:52 GMT"   
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "X-Ms-Correlation-Request-Id": "5b82009f-eebc-4022-b4ab-980cfcd2269b" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/oauth2/token?scope=repository%3Aartifactstest%3Apull%2Cpush&service=xxx.azurecr.io" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:23:52 GMT"   
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains" 
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "X-Ms-Correlation-Request-Id": "ba574568-f2d0-4c1e-b074-6b65a72b39d8" 
DEBU[0000]    "X-Ms-Ratelimit-Remaining-Calls-Per-Second": "166.65" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Accept": "application/vnd.oci.image.index.v1+json" 
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Etag": "\"sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b\"" 
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "Docker-Content-Digest": "sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000]    "X-Ms-Correlation-Request-Id": "3236f1f8-061e-433a-896f-3b68cb2859b6" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:23:52 GMT"   
DEBU[0000]    "Content-Length": "3138"                  
DEBU[0000]    "X-Ms-Client-Request-Id": ""              
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "X-Ms-Request-Id": "bc80b544-42f1-47d9-b3fd-88e32c30fcd6" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Content-Type": "application/vnd.oci.image.index.v1+json" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/testfororas5" 
DEBU[0000] > Request method: "PUT"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Content-Type": "application/vnd.oci.image.index.v1+json" 
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0001] < Response Status: "201 Created"             
DEBU[0001] < Response headers:                          
DEBU[0001]    "Docker-Content-Digest": "sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0001]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0001]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0001]    "X-Content-Type-Options": "nosniff"       
DEBU[0001]    "Date": "Tue, 14 Feb 2023 10:23:52 GMT"   
DEBU[0001]    "Location": "/v2/artifactstest/manifests/sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0001]    "X-Ms-Client-Request-Id": ""              
DEBU[0001]    "X-Ms-Correlation-Request-Id": "026416ab-8329-4186-a1a3-c0c97e20ef8d" 
DEBU[0001]    "X-Ms-Request-Id": "f25d0678-31f5-4e8a-83eb-8ba6302252a0" 
DEBU[0001]    "Server": "openresty"                     
DEBU[0001]    "Content-Length": "0"                     
DEBU[0001]    "Connection": "keep-alive"                
DEBU[0001]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
Copied [oci-layout] ./local/oras:v1.0.0-rc.1 => [registry] xxx.azurecr.io/artifactstest:testfororas5
Digest: sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b
```

But now if I do 

```
oras discover --debug xxx.azurecr.io/artifactstest:testfororas5 --platform linux/s390x
```

I have this: 

```
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/testfororas5" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "401 Unauthorized"        
DEBU[0000] < Response headers:                          
DEBU[0000]    "Content-Length": "207"                   
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "Www-Authenticate": "Bearer realm=\"https://xxx.azurecr.io/oauth2/token\",service=\"xxx.azurecr.io\",scope=\"repository:artifactstest:pull\"" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:26:01 GMT"   
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "X-Ms-Correlation-Request-Id": "500f9976-b72a-47be-8de3-9b79f5e5afe5" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/oauth2/token?scope=repository%3Aartifactstest%3Apull&service=xxx.azurecr.io" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "X-Ms-Ratelimit-Remaining-Calls-Per-Second": "166.65" 
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:26:01 GMT"   
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "X-Ms-Correlation-Request-Id": "c481e3b5-28b7-4dbf-ae75-9f0faac8dbfb" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/testfororas5" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000]    "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:26:01 GMT"   
DEBU[0000]    "X-Ms-Request-Id": "d5cc536c-5acf-4db8-bf13-780ed6d445ea" 
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Docker-Content-Digest": "sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000]    "Etag": "\"sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b\"" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "Content-Length": "3138"                  
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "X-Ms-Client-Request-Id": ""              
DEBU[0000]    "X-Ms-Correlation-Request-Id": "92bc2f8a-e9d3-42c2-9fa8-fb36837f375f" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "Content-Type": "application/vnd.oci.image.index.v1+json" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/referrers/sha256:d1903141acefdeed52e34811d495f7569aba7b97d69c5a7ad22a3f213679b117" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "X-Ms-Correlation-Request-Id": "e6790bbc-38ce-4e94-86ca-2773bc9f1dd7" 
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:26:01 GMT"   
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "Content-Type": "application/vnd.oci.image.index.v1+json; charset=utf-8" 
DEBU[0000]    "Content-Length": "89"                    
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
Discovered 0 artifact referencing testfororas5
Digest: sha256:d1903141acefdeed52e34811d495f7569aba7b97d69c5a7ad22a3f213679b117
```
But if I do 

```
oras discover --debug xxx.azurecr.io/artifactstest:testfororas5
```

```
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/testfororas5" 
DEBU[0000] > Request method: "HEAD"                     
DEBU[0000] > Request headers:                           
DEBU[0000]    "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "401 Unauthorized"        
DEBU[0000] < Response headers:                          
DEBU[0000]    "Content-Length": "207"                   
DEBU[0000]    "X-Ms-Correlation-Request-Id": "deb34173-c307-4269-b59a-4a942d2bd3b5" 
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "Www-Authenticate": "Bearer realm=\"https://xxx.azurecr.io/oauth2/token\",service=\"xxx.azurecr.io\",scope=\"repository:artifactstest:pull\"" 
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:27:25 GMT"   
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/oauth2/token?scope=repository%3Aartifactstest%3Apull&service=xxx.azurecr.io" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:27:25 GMT"   
DEBU[0000]    "Content-Type": "application/json; charset=utf-8" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "X-Ms-Correlation-Request-Id": "e6238594-988e-4f9a-b2d3-97478e1d4ed4" 
DEBU[0000]    "X-Ms-Ratelimit-Remaining-Calls-Per-Second": "166.65" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains" 
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/manifests/testfororas5" 
DEBU[0000] > Request method: "HEAD"                     
DEBU[0000] > Request headers:                           
DEBU[0000]    "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "X-Ms-Client-Request-Id": ""              
DEBU[0000]    "Content-Type": "application/vnd.oci.image.index.v1+json" 
DEBU[0000]    "Content-Length": "3138"                  
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "X-Ms-Correlation-Request-Id": "4446a9a2-adf7-48eb-bac9-49e05e4d9e20" 
DEBU[0000]    "X-Ms-Request-Id": "30574f2a-e5fd-489d-b22c-3628a2d6f997" 
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:27:26 GMT"   
DEBU[0000]    "Docker-Content-Digest": "sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "Etag": "\"sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b\"" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000] > Request URL: "https://xxx.azurecr.io/v2/artifactstest/referrers/sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b" 
DEBU[0000] > Request method: "GET"                      
DEBU[0000] > Request headers:                           
DEBU[0000]    "Authorization": "*****"                  
DEBU[0000]    "User-Agent": "oras/1.0.0-rc.1+unreleased" 
DEBU[0000] < Response Status: "200 OK"                  
DEBU[0000] < Response headers:                          
DEBU[0000]    "Content-Type": "application/vnd.oci.image.index.v1+json; charset=utf-8" 
DEBU[0000]    "Docker-Distribution-Api-Version": "registry/2.0" 
DEBU[0000]    "Access-Control-Expose-Headers": "Docker-Content-Digest, WWW-Authenticate, Link, X-Ms-Correlation-Request-Id" 
DEBU[0000]    "Strict-Transport-Security": "max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains" 
DEBU[0000]    "X-Content-Type-Options": "nosniff"       
DEBU[0000]    "X-Ms-Correlation-Request-Id": "c62cbacd-9b37-46ec-b8a7-531db5992eb3" 
DEBU[0000]    "Date": "Tue, 14 Feb 2023 10:27:26 GMT"   
DEBU[0000]    "Connection": "keep-alive"                
DEBU[0000]    "Server": "openresty"                     
DEBU[0000]    "Content-Length": "632"                   
Discovered 2 artifacts referencing testfororas5
Digest: sha256:4169ef297659f53e8660989efb6793578b14817982078592801d1522927f511b

Artifact Type           Digest
application/spdx+sbom   sha256:d9cf57f84c4f1fcb51e939915f100c14fee8c60111e1d0d2ea5d160d9a8b42f7
application/spdx+json   sha256:4fe1c210244e0c27524ef9b8d28463cb8d5d093a30615ecaf4c7d1c6a92e2908
```

So right now I don't know on what these two artifacts are related and it's weird that no artifact is related to  the linux/s390x platform

### What did you expect to happen?

Normally when I don't use the platform flag I will have no discovered artefact. But with the platform flag, I need to have 1 artefact named sbom.

### How can we reproduce it?

Ubuntu jammy and theses commands:

```
oras cp ghcr.io/oras-project/oras:v1.0.0-rc.1 --to-oci-layout ./local/oras:v1.0.0-rc.1 -v
oras attach --debug --oci-layout --artifact-type test.sbom ./local/oras@sha256:d1903141acefdeed52e34811d495f7569aba7b97d69c5a7ad22a3f213679b117 sbom 
oras discover --oci-layout ./local/oras:v1.0.0-rc.1 -v -o tree --platform linux/s390x
oras cp -r --debug --from-oci-layout -v ./local/oras:v1.0.0-rc.1 xxx.azurecr.io/artifactstest:testfororas
oras discover --debug xxx.azurecr.io/artifactstest:testfororas5 --platform linux/s390x
oras discover --debug xxx.azurecr.io/artifactstest:testfororas5
```

### What is the version of your ORAS CLI?

Version:        1.0.0-rc.1+unreleased
Go version:     go1.20
Git commit:     8bda262f901e8ef8404cf23d59c5f47f4280fdf3
Git tree state: clean

### What is your OS environment?

Ubuntu jammy

### Are you willing to submit PRs to fix it?

- [ ] Yes, I am willing to fix it.

### PR DESCRIPTION
**Pull request by qweeah**
This PR fixes the bug in digest printing and aligns the copy behavior of OCI image layout and remote registry.

Related E2E test specs are also updated.

Resolves #812, #826

### CODE REVIEW

File: cmd/oras/cp.go [hunk 1]
```diff
@@ -117,6 +120,9 @@ func runCopy(opts copyOptions) error {
 	committed := &sync.Map{}
 	extendedCopyOptions := oras.DefaultExtendedCopyOptions
 	extendedCopyOptions.Concurrency = opts.concurrency
+	extendedCopyOptions.FindPredecessors = func(ctx context.Context, src content.ReadOnlyGraphStorage, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		return graph.Referrers(ctx, src, desc, "")
```
Comment (shizhMSFT) [inline cmd/oras/cp.go:124]: Using `graph.Referrers` is workable but not efficient as we don't need to fetch the manifest for artifact types and annotations, which we don't care here.

File: cmd/oras/cp.go [hunk 2]
```diff
@@ -131,34 +137,43 @@ func runCopy(opts copyOptions) error {
...
 			err = oras.ExtendedCopyGraph(ctx, src, dst, desc, extendedCopyOptions.ExtendedCopyGraphOptions)
 		} else {
 			err = oras.CopyGraph(ctx, src, dst, desc, extendedCopyOptions.CopyGraphOptions)
 		}
+		if err != nil {
+			return err
+		}
+		if ref != "" {
+			err = dst.Tag(ctx, desc, ref)
+		}
```
Comment (shizhMSFT) [inline cmd/oras/cp.go:159]: By decomposing `ExtendedCopy` to `Resolve`-`ExtendedCopyGraph`-`Tag`, we lose the optimization for `FetchReference` and `PushReference` when platform is not specified, which is the most case.

File: cmd/oras/cp.go (hunk 1 above)
Comment (qweeah) [reply cmd/oras/cp.go:124]: Shouldn't remote repository always be a `ReferrerLister` and don't need to enrich descriptor?

File: cmd/oras/cp.go (hunk 2 above)
Comment (qweeah) [reply cmd/oras/cp.go]: Yes, you are right. Updated to only do `Resolve`-`ExtendedCopyGraph`-`Tag` when recursively copying with platform specified.

File: cmd/oras/cp.go [hunk 3]
```diff
@@ -131,17 +147,26 @@ func runCopy(opts copyOptions) error {
...
 	} else {
 		if opts.recursive {
 			desc, err = oras.ExtendedCopy(ctx, src, opts.From.Reference, dst, opts.To.Reference, extendedCopyOptions)
```
Comment (shizhMSFT) [inline cmd/oras/cp.go:159]: nit:
```suggestion
			desc, err = oras.ExtendedCopy(ctx, src, opts.From.Reference, dst, ref, extendedCopyOptions)
```

File: cmd/oras/cp.go [hunk 4]
```diff
@@ -131,17 +147,26 @@ func runCopy(opts copyOptions) error {
...
+	if ref := opts.To.Reference; ref == "" || (opts.recursive && opts.Platform.Platform != nil) {
+		// need resolving first
+		rOpts := oras.DefaultResolveOptions
+		rOpts.TargetPlatform = opts.Platform.Platform
+		desc, err = oras.Resolve(ctx, src, opts.From.Reference, rOpts)
 		if err != nil {
 			return err
 		}
 		if opts.recursive {
+			// not ExtendCopy since we already have desc
```
Comment (shizhMSFT) [inline cmd/oras/cp.go:146]: nit: this comment can be removed.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**